quic/s2n-quic-core/src/crypto/tls/testing.rs

```diff
@@ -149,12 +154,26 @@ impl<S: tls::Session, C: tls::Session> Pair<S, C> {
         self.iterations += 1;
 
         eprintln!("1/2 RTT");
+        self.check_progress(client_hello_cb_done);
 
-        self.check_progress();
         Ok(())
     }
 
-    fn check_progress(&self) {
+    fn check_progress(&mut self, client_hello_cb_done: Option<Arc<AtomicBool>>) {
+        if let Some(client_hello_cb_done) = client_hello_cb_done {
+            // The 2nd iteration is when the server receives and processes the ClientHello. If
+            // the client_hello callback is not done, then the server is not done processing the
+            // client hello and can't make progress. So return early and wait for the
+            // client_hello callback to complete first.
+            if !client_hello_cb_done.load(Ordering::SeqCst) {
+                // reset the iteration and wait for client hello callback to complete
+                if self.iterations == 2 {
+                    self.iterations -= 1;
+                    return;
+                }
```

Comment:
would it make sense to check if the client_hello_cb_done before iterations is incremented on line 154?

Reply:
This code path is also shared by tests that dont set the callback. So i wanted the callback specific logic to be contained to as small of a surface area as possible and make non-callback test easy to reason about.

Since the `check_progress` is where we make assertions (poll remain clean of branches), i decided to move the logic here and do a decrement instead of conditional increment.